---
title: |
  | PD Differential Gene Expression - Gene enrichment analysis
  | Hardy project dataset
subtitle: "Corrected identification script"
author: 
- name: "Guillermo Rocamora PÃ©rez"
  affiliation: UCL - Version 1.3
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  bookdown::html_document2:
    figure_caption: yes
    code_folding: show
    theme: paper
    highlight: haddock
    df_print: paged
    toc: true
    toc depth: 3
    toc_float: true
    number_sections: true
  md_document:
    variant: markdown_github
    toc: true
    number_sections: true
always_allow_html: true
link-citations: true
notes-after-punctuation: false
csl: https://raw.githubusercontent.com/citation-style-language/styles/master/dependent/jhep-reports.csl
---

```{r setup, echo = T, warning = F, message = F}
# Load libraries
library(DESeq2)
library(edgeR)
library(limma)
library(variancePartition)
library(foreach)
library(clusterProfiler)
library(org.Hs.eg.db)
library(ggh4x)
library(magrittr)
library(enrichplot)
library(patchwork)
library(fgsea)
library(doParallel)
library(tidyverse)

# Define main path
# main_path <- file.path("/home/grocamora/RytenLab-Research/18-DGE_limma_dream")# here::here()

main_path <- file.path("/home/jbrenton/ASAP_Hardy_post_align_files/Final_Dataset/Hardy_Limma_Analysis_Final_version_only")

# Load helper functions
# source(file.path(main_path, "R/hf_additional.R"))
source(file.path(main_path, "R/hf_graph_and_themes.R"))
# source(file.path(main_path, "R/hf_intron_categories.R"))
source(file.path(main_path, "R/hf_reports.R"))

# Set options
options(dplyr.summarise.inform = FALSE)
options(lifecycle_verbosity = "warning")
knitr::opts_chunk$set(echo = T, warning = F, message = F, 
                      out.width="100%", fig.align = "center", dpi = 150,
                      fig.width = 7.2, fig.height = 7.2)
```

```{r setup-theme, echo = F, results = "asis"}
## Custom sciRmdTheme with increased width.
## Please ask guillermorocamora@gmail.com for more details.
## Download from devtools::install_github("https://github.com/guillermo1996/grpSciRmdTheme")
if(require(grpSciRmdTheme)){
  grpSciRmdTheme::set.theme(
    theme = "default",
    color = NULL,
    header.sticky = FALSE,
    list.group.icon = "arrow",
    font.family = "Arial",
    font.color = "black",
    header.color = "darkblue",
    figure.perc = 100
  )
}

## Custom styles for the output html
cat('
<style type="text/css">
.dataTables_scrollHeadInner{
  width:100% !important;
}
.dataTables_scrollHeadInner table{
  width:100% !important;
}
/*
.code-folding-btn {
  display: none;
}
*/
h3, .h3 {
  font-size: 22px!important;
}
h4, .h4 {
  font-size: 18px!important;
}
h5, .h5 {
  font-size: 16px!important;
}
body{
  font-size: 13px;
}

.tocify-subheader {
    text-indent: 15px;
    display: none;
    font-size: 12px;
}

.break5 .nav-tabs, .break4 .nav-tabs, .break3 .nav-tabs, .break2 .nav-tabs{
  display: grid;
  text-align: center;
  font-size: 14px;
  font-weight: bold;
}
.break5 .nav-tabs:before, .break4 .nav-tabs:before, .break3 .nav-tabs:before, .break2 .nav-tabs:before,
.break5 .nav-tabs:after, .break4 .nav-tabs:after, .break3 .nav-tabs:after, .break2 .nav-tabs:after {
  content: unset;
}

.break5 .nav-tabs{
  grid-template-columns: repeat(5, 1fr);
}
.break4 .nav-tabs{
  grid-template-columns: repeat(4, 1fr);
}
.break3 .nav-tabs{
  grid-template-columns: repeat(3, 1fr);
}
.break2 .nav-tabs{
  grid-template-columns: repeat(2, 1fr);
}

.breakB5 .nav-tabs{
  grid-template-columns: repeat(5, 1fr);
  font-size: 12px;
}
.breakB4 .nav-tabs{
  grid-template-columns: repeat(4, 1fr);
  font-size: 12px;
}
.breakB3 .nav-tabs{
  grid-template-columns: repeat(3, 1fr);
  font-size: 12px;
}
.breakB2 .nav-tabs{
  grid-template-columns: repeat(2, 1fr);
  font-size: 12px;
}

.stop-break .nav-tabs {
  display: flex;
  font-size: 13px;
  font-weight: normal;
}
.stop-break .nav-tabs:before,
.stop-break .nav-tabs:after {
  content: unset;
}
</style>')
```

# Background

**Methodoly and documentation: TBD**

This report includes the gene enrichment analysis of the differentially expressed genes found in the previous report. We decided to focus on three different databases: *GO*, *KEGG* and *REACTOME*. 

First, we need to load the results from the previous report and set the parameters that will dictate the results:

```{r setup-variables}
# Paths
# main_path <- "/home/grocamora/RytenLab-Research/18-DGE_limma_dream"
# results_path <- file.path(main_path, "results/Hardy_NewCovs")
results_path <- file.path(main_path, "results")

hardy_dge_results_path <- file.path(results_path, "dge_variables/")
deg_all_output_path <- file.path(hardy_dge_results_path, "PD_Collapsed_dge_all.rds")
gene_enrichment_results_path <- file.path(results_path, "gene_enrichment_variables/")
dir.create(gene_enrichment_results_path, showWarnings = F, recursive = T)

formula_res_merge_path <- file.path(gene_enrichment_results_path, "pd_collapsed_formula_GO.rds")
formula_res_merge_red_path <- file.path(gene_enrichment_results_path, "pd_collapsed_formula_GO_red.rds")
formula_res_KEGG_path <- file.path(gene_enrichment_results_path, "pd_collapsed_formula_KEGG.rds")
formula_res_REACTOME_path <- file.path(gene_enrichment_results_path, "pd_collapsed_formula_REACTOME.rds")

# Load data
deg_all_output <- readRDS(deg_all_output_path)

# Report parameters
count_threshold <- 2
go_reduce_threshold <- 0.5
```

We will employ `clusterProfiler` for the enrichment analysis. We also need to load the gene information - i.e. gene symbol, *ENTREZID* and *ENSEMBL ID*.

```{r load-gene-info}
# Load gene info
uniKeys <- AnnotationDbi::keys(org.Hs.eg.db, keytype = "ENSEMBL")
cols <- c("SYMBOL", "ENTREZID", "ENSEMBL")
cross_comp_list <- AnnotationDbi::select(org.Hs.eg.db, keys = uniKeys, columns = cols, keytype = "ENSEMBL")

# Remove duplicate symbols
non_ducplicates_idx <- which(duplicated(cross_comp_list$SYMBOL) == F)
cross_comp_list <- cross_comp_list[non_ducplicates_idx, ]
```

From the results previously loaded, we generate the clusters of data. Two different grouping scenearios are going to be considered: collapsed by brain region and brain region specific.

```{r generate-clusters}
# Generate the cluster data combined
 cluster_data <- deg_all_output %>%
  # dplyr::filter(method == "limma") %>%
  # dplyr::select(-method) %>%
  dplyr::mutate(test = stringr::str_replace_all(test, "-", " vs ")) %>% 
  dplyr::mutate(test = factor(test, levels = c("PD vs Control"))) %>%
  dplyr::left_join(cross_comp_list, by = c("gene_id" = "ENSEMBL"), relationship = "many-to-many") %>%
  dplyr::select(ENTREZID, SYMBOL, logFC, adj.P.Val, tissue, sig_reg, test) %>%
  dplyr::filter(!is.na(ENTREZID))

cluster_data_sig <- cluster_data %>%
  dplyr::filter(sig_reg != "Unchanged") %>%
  dplyr::distinct(ENTREZID, tissue, test, .keep_all = T)
universe_genes <- cluster_data$ENTREZID %>% unique()
```

## Changelog {.tabset -}

### v1.3 {-}

* Updated to R v4.3.0

* The GO reduction is now consistent with the Wood dataset.

### v1.2 {-}

* Fixed covariate identification mistake.

### v1.1 {-}

* Modified plots to better represent the pathways.

* Refactoring code to better handle the different scenarios.

### v1.0 {-}

* Initial report.

* Added GO term reduction.

## Helper functions

```{r}
generatePlotDataframe <- function(formula_res){
  if("parent_term" %in% colnames(formula_res@compareClusterResult)){
    description_names <- formula_res@compareClusterResult %>% 
      dplyr::distinct(Description, parent_term) %>% 
      dplyr::group_by(parent_term) %>% 
      dplyr::count() %>%
      dplyr::mutate(mod_description = stringr::str_to_title(paste0(parent_term, " (", n, ")"))) %>%
      dplyr::mutate(mod_description = paste0(strwrap(mod_description, width = 50, prefix = "\n", initial = ""),
                                             collapse = "")) %>%
      dplyr::select(parent_term, n, mod_description)
    
    plot_df <- formula_res@compareClusterResult %>% 
      dplyr::group_by(test, tissue, sig_reg, parent_term) %>%
      dplyr::filter(p.adjust == min(p.adjust)) %>%
      dplyr::filter(parent_sim_score == max(parent_sim_score)) %>%
      dplyr::ungroup() %>%
      dplyr::left_join(description_names, by = "parent_term") %>%
      dplyr::mutate(Description = mod_description) %>%
      # tidyr::separate(GeneRatio, into = c("gr1", "gr2"), convert = T) %>%
      # dplyr::mutate(geneRatio = gr1/gr2) %>%
      dplyr::group_by(Description) %>% 
      dplyr::mutate(min.padj = min(p.adjust)) %>% 
      dplyr::ungroup() %>%
      dplyr::arrange(desc(n), test, tissue, p.adjust) %>%
      dplyr::mutate(Description = factor(Description, levels = unique(Description)),
                    sig_reg = factor(sig_reg, levels = c("Significantly Downregulated", "Significantly Upregulated")),
                    test = factor(test, levels = c("PD vs Control")),
                    tissue = factor(tissue, levels = c("ACG", "IPL", "MFG", "MTG")))
  }else{
    plot_df <- formula_res@compareClusterResult %>% 
      dplyr::rowwise() %>% 
      dplyr::mutate(Description = stringr::str_to_title(Description),
                    Description = paste0(strwrap(Description, width = 50, prefix = "\n", initial = ""), 
                                         collapse = "")) %>% 
      dplyr::group_by(Description) %>% 
      dplyr::mutate(n = n(),
                    min.padj = min(p.adjust)) %>% 
      dplyr::ungroup() %>%
      dplyr::arrange(desc(n), min.padj, test, tissue, p.adjust) %>%
      dplyr::mutate(Description = factor(Description, levels = unique(Description)),
                    sig_reg = factor(sig_reg, levels = c("Significantly Downregulated", "Significantly Upregulated")),
                    test = factor(test, levels = c("PD vs Control")),
                    tissue = factor(tissue, levels = c("ACG", "IPL", "MFG", "MTG")))
  }
}

drawPathwaysGO <- function(df_plot, x = "test", x_expansion = c(0.5, 0.5), drop_facet = T, drop_x = F){
  plot <- plot_df %>% 
    ggplot(aes(x = .data[[x]], y = Description, fill = -log10(p.adjust))) +
    geom_tile(color = "black", width = 1, show.legend = T) +
    geom_hline(yintercept = seq(0.5, length(plot_df$Description)), color="black", linewidth = 0.2) +
    geom_vline(xintercept = seq(0.5, length(plot_df[, x, T])+1), color="black", linewidth = 0.2) +
    scale_y_discrete(limits = rev, expand = expansion()) +
    scale_x_discrete(drop = drop_x, expand = expansion(add = x_expansion)) +
    scale_fill_viridis_c(name = "-log"[1][0]~"(FDR)", breaks = c(0, 2, 4, 6, 8, 10),
                         na.value = "white", option = "cividis") +
    coord_equal() +
    labs(x = "", y = "GO Term", fill = "hola") +
    custom_gg_theme +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          strip.text.x = element_text(size = 9, face = "plain"),
          panel.background = element_rect(fill = "white"),
          panel.grid.minor = element_blank(),
          panel.grid.major.y = element_blank(),
          panel.grid.major.x = element_blank(),
          panel.grid = element_blank())
  
  # Facet the plots based on what they have on the X-axis
  if(x == "test"){
    plot <- plot +
      ggh4x::facet_nested(. ~ sig_reg, drop = drop_facet, 
                          labeller = labeller(sig_reg = c("Significantly Upregulated" = "Up-Reg.",
                                                          "Significantly Downregulated" = "Down-Reg.")))
  }else if(x == "tissue"){
    plot <- plot +
      ggh4x::facet_nested(. ~ sig_reg + test, drop = drop_facet, 
                          labeller = labeller(sig_reg = c("Significantly Upregulated" = "Up-Reg.",
                                                          "Significantly Downregulated" = "Down-Reg.")))
  }
  
  return(plot)
}

# New function to go with updated rrvgo - problem mainly with reduce sim function - both score parameter and parent_sim_score are changed
# scores is now a character vector choice of either "size" or "uniqueness" - uniqueness is default
# parent sim score is now divided into term uniqueness and term dissimilaity or something
updated_go_reduce<-function (pathway_df, orgdb = "org.Hs.eg.db", threshold = 0.7, 
                             scores = NULL, measure = "Wang") 
{
  if (!measure %in% c("Resnik", "Lin", "Rel", "Jiang", "Wang")) {
    stop("Chosen measure is not one of the recognised measures, c(\"Resnik\", \"Lin\", \"Rel\", \"Jiang\", \"Wang\").")
  }
  if (measure == "Wang") {
    computeIC <- FALSE
  }
  else {
    computeIC <- TRUE
  }
  ont <- pathway_df %>% .[["go_type"]] %>% unique()
  if (any(!ont %in% c("BP", "CC", "MF"))) {
    stop("Column go_type does not contain the recognised sub-ontologies, c(\"BP\", \"CC\", \"MF\")")
  }
  go_similarity <- setNames(object = vector(mode = "list", 
                                            length = length(ont)), nm = ont)
  for (i in 1:length(ont)) {
    print(stringr::str_c("Reducing sub-ontology: ", ont[i]))
    hsGO <- GOSemSim::godata(OrgDb = orgdb, ont = ont[i], 
                             computeIC = computeIC)
    terms <- pathway_df %>% dplyr::filter(.data$go_type == 
                                            ont[i]) %>% .[["go_id"]] %>% unique()
    sim <- GOSemSim::mgoSim(GO1 = terms, GO2 = terms, semData = hsGO, 
                            measure = measure, combine = NULL)
    go_similarity[[i]] <- rrvgo::reduceSimMatrix(simMatrix = sim, 
                                                 threshold = threshold, orgdb = orgdb) %>% 
      tibble::as_tibble() %>% dplyr::rename(parent_id = .data$parent, 
                                            parent_term = .data$parentTerm, parent_sim_score = .data$termUniqueness)
  }
  go_sim_df <- go_similarity %>% qdapTools::list_df2df(col1 = "go_type")
  pathway_go_sim_df <- pathway_df %>% 
    dplyr::inner_join(go_sim_df %>% 
                        dplyr::select(.data$go_type, go_id = .data$go, contains("parent")), 
                      by = c("go_type", "go_id")) %>% 
    dplyr::arrange(.data$go_type, .data$parent_id, -.data$parent_sim_score)
  
  return(pathway_go_sim_df)
}
```

# GO enrichment analysis {.tabset .break2}

Two different visualizations are presented: i) GO terms as extracted from `enrichGO()` and ii) GO terms collapsed by similarity from Regina H. Reynolds' function [`go_reduce()`](https://rhreynolds.github.io/rutils/reference/go_reduce.html), from her package [`rutils`](https://rhreynolds.github.io/rutils/index.html).

If several GO terms are combined into one, the *GeneRatio* and *adjusted p-value* shown in the plots correspond to the minimum *adjusted p-value* of the collapsed terms. The number in brackets next to the Y-axis labels represent the number of GO terms collapsed.

```{r}
if(!file_test("-f", formula_res_merge_path, formula_res_merge_red_path)){
  formula_res_mf <- clusterProfiler::compareCluster(
    ENTREZID ~ sig_reg + tissue + test,
    data = cluster_data_sig, fun = enrichGO,
    OrgDb = org.Hs.eg.db, ont = "MF",
    universe = universe_genes,
    readable = T) %>%
    dplyr::mutate(go_type = "MF", go_id = ID)
  
  formula_res_bp <- clusterProfiler::compareCluster(
    ENTREZID ~ sig_reg + tissue + test,
    data = cluster_data_sig, fun = enrichGO,
    OrgDb = org.Hs.eg.db, ont = "BP",
    universe = universe_genes,
    readable = T) %>%
    dplyr::mutate(go_type = "BP", go_id = ID)
  
  formula_res_cc <- clusterProfiler::compareCluster(
    ENTREZID ~ sig_reg + tissue + test,
    data = cluster_data_sig, fun = enrichGO,
    OrgDb = org.Hs.eg.db, ont = "CC",
    universe = universe_genes,
    readable = T) %>%
    dplyr::mutate(go_type = "CC", go_id = ID)
  
  # Reduce the GO terms
  formula_res_mf_red <- formula_res_mf
  formula_res_bp_red <- formula_res_bp
  formula_res_cc_red <- formula_res_cc
  
  formula_res_mf_red@compareClusterResult %<>% updated_go_reduce(threshold = go_reduce_threshold)
  formula_res_bp_red@compareClusterResult %<>% updated_go_reduce(threshold = go_reduce_threshold)
  formula_res_cc_red@compareClusterResult %<>% updated_go_reduce(threshold = go_reduce_threshold)
  
  # Filter pathways
  
  formula_res_mf@compareClusterResult %<>% dplyr::filter(Count >= count_threshold)
  formula_res_bp@compareClusterResult %<>% dplyr::filter(Count >= count_threshold)
  formula_res_cc@compareClusterResult %<>% dplyr::filter(Count >= count_threshold)
  

  formula_res_mf_red@compareClusterResult %<>% dplyr::filter(Count >= count_threshold)
  formula_res_bp_red@compareClusterResult %<>% dplyr::filter(Count >= count_threshold)
  formula_res_cc_red@compareClusterResult %<>% dplyr::filter(Count >= count_threshold)
  
  # Merge ontologies
  formula_res_merge <- clusterProfiler::merge_result(list(mf = formula_res_mf,
                                                          bp = formula_res_bp,
                                                          cc = formula_res_cc))
  formula_res_merge_red <- clusterProfiler::merge_result(list(mf = formula_res_mf_red,
                                                              bp = formula_res_bp_red,
                                                              cc = formula_res_cc_red))
  
  # Rename ontologies
  names(formula_res_merge@compareClusterResult)[1] <- "Ontology"
  names(formula_res_merge_red@compareClusterResult)[1] <- "Ontology"
  
  # Save results
  formula_res_merge %>% saveRDS(formula_res_merge_path)
  formula_res_merge_red %>% saveRDS(formula_res_merge_red_path)
}else{
  formula_res_merge <- readRDS(formula_res_merge_path)
  formula_res_merge_red <- readRDS(formula_res_merge_red_path)
}


```

## Collapsed brain regions {.tabset .breakB3}

### Both {-}

<details>
<summary>Expand to see results for default GO terms.</summary>
```{r, fig.width=10, fig.height=15, class.output="bg-warning"}
formula_res_merge_both <- formula_res_merge %>% dplyr::filter(tissue == "Collapse")
plot_df <- generatePlotDataframe(formula_res_merge_both)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, drop_facet = F)
}else{
  print("No pathways found in this scenario.")
}

plot_df$tissue<-"Collapse"
plot_df_collapse<-plot_df
```
</details>\

<details>
<summary>Expand to see results for reduced GO terms.</summary>
```{r, fig.width=7, fig.height=9, class.output="bg-warning"}
formula_res_merge_red_both <- formula_res_merge_red %>% dplyr::filter(tissue == "Collapse")
plot_df <- generatePlotDataframe(formula_res_merge_red_both)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, drop_facet = F)
}else{
  print("No pathways found in this scenario.")
}

plot_df$tissue<-"Collapse"
plot_df_collapse_red<-plot_df
```
</details>\

### Upregulated {-}

<details>
<summary>Expand to see results for default GO terms.</summary>
```{r, fig.width=10, fig.height=15, class.output="bg-warning"}
formula_res_merge_up <- formula_res_merge %>% dplyr::filter(grepl("Upregulated", sig_reg), tissue == "Collapse")
plot_df <- generatePlotDataframe(formula_res_merge_up)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df)
}else{
  print("No pathways found in this scenario.")
}

```
</details>\

<details>
<summary>Expand to see results for reduced GO terms.</summary>
```{r, fig.width=10, fig.height=15, class.output="bg-warning"}
formula_res_merge_red_up <- formula_res_merge_red %>% dplyr::filter(grepl("Upregulated", sig_reg), tissue == "Collapse")
plot_df <- generatePlotDataframe(formula_res_merge_red_up)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df)
}else{
  print("No pathways found in this scenario.")
}

```
</details>\

### Downregulated {-}

<details>
<summary>Expand to see results for default GO terms.</summary>
```{r, fig.width=10, fig.height=15, class.output="bg-warning"}
formula_res_merge_down <- formula_res_merge %>% dplyr::filter(grepl("Downregulated", sig_reg), tissue == "Collapse")
plot_df <- generatePlotDataframe(formula_res_merge_down)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df)
}else{
  print("No pathways found in this scenario.")
}
```
</details>\

<details>
<summary>Expand to see results for reduced GO terms.</summary>
```{r, fig.width=7, fig.height=9, class.output="bg-warning"}
formula_res_merge_red_down <- formula_res_merge_red %>% dplyr::filter(grepl("Downregulated", sig_reg), tissue == "Collapse")
plot_df <- generatePlotDataframe(formula_res_merge_red_down)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df)
}else{
  print("No pathways found in this scenario.")
}
```
</details>\


## By brain regions {.tabset .breakB3}

### Both {-}

<details>
<summary>Expand to see results for default GO terms.</summary>
```{r, fig.width=10, fig.height=30, class.output="bg-warning"}
formula_res_merge_both <- formula_res_merge %>% dplyr::filter(tissue != "Collapse")
plot_df <- generatePlotDataframe(formula_res_merge_both)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, x = "tissue", drop_facet = T)
}else{
  print("No pathways found in this scenario.")
}

plot_df_indv_bas<-plot_df

```


</details>\

<details>
<summary>Expand to see results for reduced GO terms.</summary>
```{r, fig.width=13, fig.height=17, class.output="bg-warning"}
formula_res_merge_red_both <- formula_res_merge_red %>% dplyr::filter(tissue != "Collapse")
plot_df <- generatePlotDataframe(formula_res_merge_red_both)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, x = "tissue", drop_facet = T) + theme(axis.text.y = element_text(size=12))
}else{
  print("No pathways found in this scenario.")
}

plot_df_indv_bas_red<-plot_df

```
</details>\



### Upregulated {-}

<details>
<summary>Expand to see results for default GO terms.</summary>
```{r, fig.width=7, fig.height=10, class.output="bg-warning"}
formula_res_merge_up <- formula_res_merge %>% dplyr::filter(grepl("Upregulated", sig_reg), tissue != "Collapse")
plot_df <- generatePlotDataframe(formula_res_merge_up)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, x = "tissue")
}else{
  print("No pathways found in this scenario.")
}
```
</details>\

<details>
<summary>Expand to see results for reduced GO terms.</summary>
```{r, fig.width=7, fig.height=5, class.output="bg-warning"}
formula_res_merge_red_up <- formula_res_merge_red %>% dplyr::filter(grepl("Upregulated", sig_reg), tissue != "Collapse")
plot_df <- generatePlotDataframe(formula_res_merge_red_up)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, x = "tissue")
}else{
  print("No pathways found in this scenario.")
}
```
</details>\

### Downregulated {-}

<details>
<summary>Expand to see results for default GO terms.</summary>
```{r, fig.width=10, fig.height=20, class.output="bg-warning"}
formula_res_merge_down <- formula_res_merge %>% dplyr::filter(grepl("Downregulated", sig_reg), tissue != "Collapse")
plot_df <- generatePlotDataframe(formula_res_merge_down)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, x = "tissue")
}else{
  print("No pathways found in this scenario.")
}

```
</details>\

<details>
<summary>Expand to see results for reduced GO terms.</summary>
```{r, fig.width=7, fig.height=12, class.output="bg-warning"}
formula_res_merge_red_down <- formula_res_merge_red %>% dplyr::filter(grepl("Downregulated", sig_reg), tissue != "Collapse")
plot_df <- generatePlotDataframe(formula_res_merge_red_down)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, x = "tissue")
}else{
  print("No pathways found in this scenario.")
}

```
</details>\


## Merge to output to csv

```{r}
go_merge_no_red<-rbind(plot_df_collapse, plot_df_indv_bas) %>% arrange(p.adjust)

write_csv(x = go_merge_no_red, file = file.path(gene_enrichment_results_path, "PD_collapsed_Sig_GO_terms_no_reduction.csv"))

go_merge_red<-rbind(plot_df_collapse_red,plot_df_indv_bas_red) %>% arrange(p.adjust)

write_csv(x = go_merge_red, file = file.path(gene_enrichment_results_path, "PD_collapsed_Sig_GO_terms_merged_with_term_reduction.csv"))

```


# KEGG enrichment analysis {.tabset .break2}

```{r}
if(!file_test("-f", formula_res_KEGG_path)){
  formula_res_KEGG <- clusterProfiler::compareCluster(
    ENTREZID ~ sig_reg + tissue + test,
    data = cluster_data_sig, fun = enrichKEGG,
    organism = "hsa",
    universe = universe_genes) %>%
    clusterProfiler::setReadable(OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
  
  formula_res_KEGG@compareClusterResult %<>% dplyr::filter(Count >= count_threshold)
  
  # Save results
  formula_res_KEGG %>% saveRDS(formula_res_KEGG_path)
}else{
  formula_res_KEGG <- readRDS(formula_res_KEGG_path)
}
```

## Collapsed brain regions {.tabset .breakB3}

### Both {-}

```{r, fig.width=6, fig.height=4, class.output="bg-warning"}
formula_res_KEGG_both <- formula_res_KEGG %>% dplyr::filter(tissue == "Collapse")
plot_df <- generatePlotDataframe(formula_res_KEGG_both)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, drop_facet = F)
}else{
  print("No pathways found in this scenario.")
}

plot_df$tissue<-"Collapse"
kegg_plot_df_collapse<- plot_df
```

### Upregulated {-}

```{r, fig.width=10, fig.height=15, class.output="bg-warning"}
formula_res_KEGG_up <- formula_res_KEGG %>% dplyr::filter(grepl("Upregulated", sig_reg), tissue == "Collapse")
plot_df <- generatePlotDataframe(formula_res_KEGG_up)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df)
}else{
  print("No pathways found in this scenario.")
}
```

### Downregulated {-}

```{r, fig.width=5, fig.height=4, class.output="bg-warning"}
formula_res_KEGG_down <- formula_res_KEGG %>% dplyr::filter(grepl("Downregulated", sig_reg), tissue == "Collapse")
plot_df <- generatePlotDataframe(formula_res_KEGG_down)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df)
}else{
  print("No pathways found in this scenario.")
}
```

## By brain regions {.tabset .breakB3}

### Both {-}

```{r, fig.width=10, fig.height=4, class.output="bg-warning"}
formula_res_KEGG_both <- formula_res_KEGG %>% dplyr::filter(tissue != "Collapse")
plot_df <- generatePlotDataframe(formula_res_KEGG_both)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, x = "tissue", drop_facet = F)
}else{
  print("No pathways found in this scenario.")
}

kegg_plot_df_indv_bas<- plot_df

```

### Upregulated {-}

```{r, fig.width=6, fig.height=4, class.output="bg-warning"}
formula_res_KEGG_up <- formula_res_KEGG %>% dplyr::filter(grepl("Upregulated", sig_reg), tissue != "Collapse")
plot_df <- generatePlotDataframe(formula_res_KEGG_up)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, x = "tissue")
}else{
  print("No pathways found in this scenario.")
}
```

### Downregulated {-}

```{r, fig.width=6, fig.height=4, class.output="bg-warning"}
formula_res_KEGG_down <- formula_res_KEGG %>% dplyr::filter(grepl("Downregulated", sig_reg), tissue != "Collapse")
plot_df <- generatePlotDataframe(formula_res_KEGG_down)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, x = "tissue")
}else{
  print("No pathways found in this scenario.")
}
```

## Merge to output to csv

```{r}
kegg_merge<-rbind(kegg_plot_df_collapse, kegg_plot_df_indv_bas) %>% arrange(p.adjust)

write_csv(x = kegg_merge, file = file.path(gene_enrichment_results_path, "PD_collapsed_Sig_KEGG_terms.csv"))
```


# REACTOME enrichment analysis {.tabset .break2}

```{r}
if(!file_test("-f", formula_res_REACTOME_path)){
  formula_res_REACTOME <- clusterProfiler::compareCluster(
    ENTREZID ~ sig_reg + tissue + test,
    data = cluster_data_sig, fun = "enrichPathway",
    readable = T)
  
  formula_res_REACTOME@compareClusterResult %<>% dplyr::filter(Count >= count_threshold)
  
  # Save results
  formula_res_REACTOME %>% saveRDS(formula_res_REACTOME_path)
}else{
  formula_res_REACTOME <- readRDS(formula_res_REACTOME_path)
}
```

## Collapsed brain regions {.tabset .breakB3}

### Both {-}

```{r, fig.width=7, fig.height=6.5, class.output="bg-warning"}
formula_res_REACTOME_both <- formula_res_REACTOME %>% dplyr::filter(tissue == "Collapse")
plot_df <- generatePlotDataframe(formula_res_REACTOME_both)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, drop_facet = F)
}else{
  print("No pathways found in this scenario.")
}

plot_df$tissue<-"Collapse"
reactome_plot_df_collapse <- plot_df
```

### Upregulated {-}

```{r, fig.width=10, fig.height=15, class.output="bg-warning"}
formula_res_REACTOME_up <- formula_res_REACTOME %>% dplyr::filter(grepl("Upregulated", sig_reg), tissue == "Collapse")
plot_df <- generatePlotDataframe(formula_res_REACTOME_up)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df)
}else{
  print("No pathways found in this scenario.")
}
```

### Downregulated {-}

```{r, fig.width=7, fig.height=6.5, class.output="bg-warning"}
formula_res_REACTOME_down <- formula_res_REACTOME %>% dplyr::filter(grepl("Downregulated", sig_reg), tissue == "Collapse")
plot_df <- generatePlotDataframe(formula_res_REACTOME_down)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df)
}else{
  print("No pathways found in this scenario.")
}
```

## By brain regions {.tabset .breakB3}

### Both {-}

```{r, fig.width=10, fig.height=6, class.output="bg-warning"}
formula_res_REACTOME_both <- formula_res_REACTOME %>% dplyr::filter(tissue != "Collapse")
plot_df <- generatePlotDataframe(formula_res_REACTOME_both)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, x = "tissue", drop_facet = F)
}else{
  print("No pathways found in this scenario.")
}

reactome_plot_df_indv_bas <- plot_df

```

### Upregulated {-}

```{r, fig.width=6, fig.height=5, class.output="bg-warning"}
formula_res_REACTOME_up <- formula_res_REACTOME %>% dplyr::filter(grepl("Upregulated", sig_reg), tissue != "Collapse")
plot_df <- generatePlotDataframe(formula_res_REACTOME_up)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, x = "tissue")
}else{
  print("No pathways found in this scenario.")
}
```

### Downregulated {-}

```{r, fig.width=6, fig.height=6, class.output="bg-warning"}
formula_res_REACTOME_down <- formula_res_REACTOME %>% dplyr::filter(grepl("Downregulated", sig_reg), tissue != "Collapse")
plot_df <- generatePlotDataframe(formula_res_REACTOME_down)

if(nrow(plot_df) > 0){
  drawPathwaysGO(plot_df, x = "tissue")
}else{
  print("No pathways found in this scenario.")
}
```

## Merge to output to csv

```{r}
reactome_merge<-rbind(reactome_plot_df_collapse,reactome_plot_df_indv_bas) %>% arrange(p.adjust)

write_csv(x = reactome_merge, file = file.path(gene_enrichment_results_path, "PD_collapsed_Sig_REACTOME_terms.csv"))
```


# Session Info

<details>

<summary>Click to Show</summary>

```{r}
sessioninfo::session_info()
```

</details>





